core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -2302,21 +2302,35 @@ class KafkaController(val config: KafkaConfig,
     }
   }
 
-  private def validateReplicas(topicPartition: TopicPartition, replicas: Option[Seq[Int]]): Option[ApiError] = {
+  private def validateReplicas(topicPartition: TopicPartition, replicas: Option[Seq[Int]]): Either[ApiError, Seq[Int]] = {
     replicas match {
-      case Some(targetReplicas) => validateTargetReplicas(topicPartition, targetReplicas)
+      case Some(targetReplicas) => {
+        validateTargetReplicas(topicPartition, targetReplicas)
+        Right(targetReplicas)
+      }
       case None => {
         // this is trying to cancel an existing replica reassignment
         val replicaAssignment = controllerContext.partitionFullReplicaAssignment(topicPartition)
         if (replicaAssignment.isBeingReassigned) {
-          val originalReplicasAlive = replicaAssignment.originReplicas.toSet.subsetOf(controllerContext.liveBrokerIds)
-          if (!originalReplicasAlive)
-            Some(new ApiError(Errors.INVALID_REPLICA_ASSIGNMENT,
-              s"Replica assignment cancellation has original brokers that are not alive. " +
-                "Replica list: " + s"${replicaAssignment.originReplicas}, live broker list: ${controllerContext.liveBrokerIds}"))
-          else None
+          val allOriginalReplicasAlive = replicaAssignment.originReplicas.toSet.subsetOf(controllerContext.liveBrokerIds)
+          if (allOriginalReplicasAlive) {
+            // Simply fallback to the original set of replicas
+            Right(replicaAssignment.originReplicas)
+          } else {
+            val deadOriginalReplicas = replicaAssignment.originReplicas.toSet.diff(controllerContext.liveBrokerIds);
+            if (deadOriginalReplicas.size >= 2) {
+              // If there are two or more dead replicas, ISR would fall below minISR, which we don't allow
```

Comment:
This is still the hard-coded logic Krit flagged, right?  This would allow cancellations with one dead broker even if it were configured at RF == minISR, and it would disallow cancellations with two dead brokers even if RF == minISR + 3 (for example). What we'd want here is a check for (deadOriginalReplicas.size > RF - minISR), right?

Reply:
Based on the discussion with Krit yesterday, a config to control this behavior would be adequate. We don’t allow topic creation with `minISR == RF` or `minISR == RF-1` because we want to be able to handle one planned and one unplanned failure all the times. So `RF - minISR >= 2` is true in all prod/ei use-cases.

Under some test scenarios, where `RF >= 5`, the config I'm going to add will allow cancellation if enough original replicas are on board. 

Does this look right to you?

Reply:
Depending on another config when we've already got two feels kludgy and error-prone to me, especially given that its description is fairly complex and subtle. Can we not simply derive this value? KafkaConfig already has min.insync.replicas and default.replication.factor; aren't those effectively what CC checks? I know individual topics can override this, and it would be better to check the RF of the topic in question explicitly, but I don't know offhand how accessible that metadata is at this point in the code.

Reply:
We've two combinations . `minISR==2, RF==4` in KAC/KRC. then we've `minISR==1, RF==3` everywhere else. so defaults in linkedin-kafka level are not the same. In fact, in linkedin-kafka `Defaults.DefaultReplicationFactor` is `1`. 

Either way, the `RF-minISR` is 2 in Linkedin clusters. I think the suggestion made was to have maximum flexibility and potentially cater to other scenarios such as dev clusters and change in future etc. Therefore, I prefer a direct control independent of the other parameters, which as you said could change per topic.

Reply:
Let me put it this way:  suppose you do this and set the global default to 2 as you've indicated, and then at some future point someone creates a cluster with RF = minISR + 1. They're not even aware of this new config, and as a result, you and they have just reintroduced the leaderless-partitions bug for that cluster. That's why it's a bad approach.

In general, _every_ manual approach is a bad approach:  if it _can_ be automated, it _should_ be automated. It's not always feasible to do so, but in this case the logic is clearcut, and the controller has total knowledge of all of the relevant pieces of info, so let's reduce the likelihood of errors rather than increase it.